### ISSUE TITLE
ReferenceError: *_1 is not defined in generic function return

### ISSUE THREAD

**Issue by mspeichert** [microsoft/TypeScript#40302]
**TypeScript Version:**  4.0.2

**Search Terms:** 
`*_1 is not defined`, generic function, compiler bug

**Code**
Hello there,
I've found some interesting bug, that seems to work only in extremely edge conditions. 
When created a simple file with only one exported function:
```ts
// g.ts
export const g = () => "hello world";
```

and the main file, that imports the function and uses it in a very specific way:
```ts
// index.ts
import { g } from "./g";

type Foo<T> = Partial<T> & { foo: number };

const f = <T, T2 extends Foo<T>>(value: T2) => ({ ...value, bar: g() });

f({ foo: 1 });
```
The result is that we get a runtime Reference Error.
Funny enough, changing almost anything in the example above results in the correct behavior.
Moreover, simply adding an invocation of `g()` at the end of the file fixes the issue.
The bug doesn't appear in version 3.8.3, but is present in version 3.9.7 and in current version as well.

**Expected behavior:** Function completes without exception.

**Actual behavior:** `ReferenceError: g_1 is not defined`

**Playground Link:** https://codesandbox.io/s/recursing-stallman-0fh49 

**Related Issues:** This issue seems like a quite an edge case, I've searched through similar bugs but none really seem to fit.

**Comment by RyanCavanaugh** [comment]
I can't repro this in `tsc`; it's possible `codesandbox.io` has a bug in it. What compiler settings are you using?

**Comment by mspeichert** [comment]
I am able to replicate the issue outside of `codesandbox.io`, but only using a bundler - parcel.js or webpack, with the same result. Also in webpack I tried two most popular typescript loaders (ts-loader and awesome-typescript-loader) and both are generating incorrect code. [Here I uploaded examples of bundled results from webpack and parcel.](https://codesandbox.io/s/frosty-heisenberg-y54gh)

I managed to do this with only default compiler options.
It might not be a compiler issue, but a bundler/loader issue, I can't really tell, but the fact that changing typescript version to 3.8.3 fixes the issue is interesting
